src/interface/swarm.cpp

```diff
@@ -317,28 +354,469 @@ void Swarm::Defrag() {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 0) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 0) = n;
+          neighborIndices_h(0, 2, 0) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 0) = n;
+        }
+      } else if (i == 0) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 1) = n;
+          neighborIndices_h(0, 0, 2) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 1) = n;
+          neighborIndices_h(0, 3, 2) = n;
+        }
+      } else if (i == 1) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 3) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 3) = n;
+          neighborIndices_h(0, 2, 3) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 3) = n;
+        }
+      }
+    } else {
+      PARTHENON_FAIL("3D particles not currently supported!");
+    }
+  }
+
+  neighborIndices_.DeepCopy(neighborIndices_h);
+}
+
+bool Swarm::Send(BoundaryCommSubset phase) {
+  auto blockIndex_h = blockIndex_.GetHostMirrorAndCopy();
+  auto mask_h = mask_.data.GetHostMirrorAndCopy();
+  auto swarm_d = GetDeviceContext();
+
+  auto pmb = GetBlockPointer();
+
+  // Fence to make sure particles aren't currently being transported locally
+  pmb->exec_space.fence();
+
+  int nbmax = vbswarm->bd_var_.nbmax;
+  ParArrayND<int> num_particles_to_send("npts", nbmax);
+  auto num_particles_to_send_h = num_particles_to_send.GetHostMirror();
+  for (int n = 0; n < nbmax; n++) {
+    num_particles_to_send_h(n) = 0;
+    auto &nb = pmb->pbval->neighbor[n];
+  }
+  int particle_size = GetParticleDataSize();
+  vbswarm->particle_size = particle_size;
+
+  int max_indices_size = 0;
+  for (int n = 0; n <= max_active_index_; n++) {
+    if (mask_h(n)) {
+      // This particle should be sent
+      if (blockIndex_h(n) >= 0) {
+        num_particles_to_send_h(blockIndex_h(n))++;
+        if (max_indices_size < num_particles_to_send_h(blockIndex_h(n))) {
+          max_indices_size = num_particles_to_send_h(blockIndex_h(n));
+        }
+      }
+    }
+  }
+  // Size-0 arrays not permitted but we don't want to short-circuit subsequent logic that
+  // indicates completed communications
+  max_indices_size = std::max<int>(1, max_indices_size);
+  // Not a ragged-right array, just for convenience
+  ParArrayND<int> particle_indices_to_send("Particle indices to send", nbmax,
+                                           max_indices_size);
+  auto particle_indices_to_send_h = particle_indices_to_send.GetHostMirror();
+  std::vector<int> counter(nbmax, 0);
+  for (int n = 0; n <= max_active_index_; n++) {
+    if (mask_h(n)) {
+      if (blockIndex_h(n) >= 0) {
+        particle_indices_to_send_h(blockIndex_h(n), counter[blockIndex_h(n)]) = n;
+        counter[blockIndex_h(n)]++;
+      }
+    }
+  }
+  num_particles_to_send.DeepCopy(num_particles_to_send_h);
+  particle_indices_to_send.DeepCopy(particle_indices_to_send_h);
```

Comment:
wait is this actually set to anything? I don't see `particle_indices_to_send_h` set?

Reply:
line 504 is `particle_indices_to_send_h(blockIndex_h(n), counter[blockIndex_h(n)]) = n;`

Reply:
Ah ok. I must have just missed that.